File: sdk-typescript/packages/scene/src/scene.ts
/**
 * WAVE SDK - Scene AI API
 *
 * AI-powered scene detection, analysis, and segmentation.
 *
 * NOTE: This is a client SDK. All authorization checks are performed server-side.
 * The API will return 403 Forbidden if the user lacks required permissions.
 */

import type {
  WaveClient,
  PaginationParams,
  PaginatedResponse,
} from '@wave-av/core';

import type {
  CreateSceneDetectionRequest,
  ListSceneDetectionsParams,
  Scene,
  SceneBoundary,
  SceneComparison,
  SceneDetection,
  SceneLabel,
  SceneType,
  Shot,
  ShotType,
} from './scene-types';

export * from './scene-types';

export class SceneAPI {
  private readonly client: WaveClient;
  private readonly basePath = '/v1/scene';

  constructor(client: WaveClient) {
    this.client = client;
  }


  /**
   * Start scene detection
   *
   * Requires: scene:detect permission
   */
  async detect(request: CreateSceneDetectionRequest): Promise<SceneDetection> {
    return this.client.post<SceneDetection>(`${this.basePath}/detect`, request);
  }

  /**
   * Get scene detection job
   *
   * Requires: scene:read permission
   */
  async getDetection(detectionId: string): Promise<SceneDetection> {
    return this.client.get<SceneDetection>(`${this.basePath}/${detectionId}`);
  }

  /**
   * Remove scene detection
   *
   * Requires: scene:remove permission (server-side RBAC enforced)
   */
  async removeDetection(detectionId: string): Promise<void> {
    await this.client.delete(
      `${this.basePath}/${detectionId}`,
      { method: 'DELETE' }
    );
  }

  /**
   * List scene detections
   *
   * Requires: scene:read permission
   */
  async listDetections(
    params?: ListSceneDetectionsParams
  ): Promise<PaginatedResponse<SceneDetection>> {
    return this.client.get<PaginatedResponse<SceneDetection>>(this.basePath, {
      params: params as unknown as Record<string, string | number | boolean | undefined>,
    });
  }


  /**
   * Get scenes for a detection
   *
   * Requires: scene:read permission
   */
  async getScenes(
    detectionId: string,
    params?: PaginationParams & {
      scene_type?: SceneType;
      min_duration?: number;
      min_confidence?: number;
    }
  ): Promise<PaginatedResponse<Scene>> {
    return this.client.get<PaginatedResponse<Scene>>(
      `${this.basePath}/${detectionId}/scenes`,
      { params: params as unknown as Record<string, string | number | boolean | undefined> }
    );
  }

  /**
   * Get a specific scene
   *
   * Requires: scene:read permission
   */
  async getScene(detectionId: string, sceneId: string): Promise<Scene> {
    return this.client.get<Scene>(
      `${this.basePath}/${detectionId}/scenes/${sceneId}`
    );
  }

  /**
   * Update scene metadata
   *
   * Requires: scene:update permission
   */
  async updateScene(
    detectionId: string,
    sceneId: string,
    updates: {
      scene_type?: SceneType;
      description?: string;
      labels?: SceneLabel[];
    }
  ): Promise<Scene> {
    return this.client.patch<Scene>(
      `${this.basePath}/${detectionId}/scenes/${sceneId}`,
      updates
    );
  }

  /**
   * Get scene at a specific timestamp
   *
   * Requires: scene:read permission
   */
  async getSceneAtTime(detectionId: string, timestamp: number): Promise<Scene | null> {
    try {
      return await this.client.get<Scene>(
        `${this.basePath}/${detectionId}/scenes/at`,
        { params: { timestamp } as unknown as Record<string, string | number | boolean | undefined> }
      );
    } catch (error) {
      if ((error as { statusCode?: number }).statusCode === 404) {
        return null;
      }
      throw error;
    }
  }


  /**
   * Get scene boundaries (transitions)
   *
   * Requires: scene:read permission
   */
  async getBoundaries(
    detectionId: string,
    params?: PaginationParams & { type?: SceneBoundary['type'] }
  ): Promise<PaginatedResponse<SceneBoundary>> {
    return this.client.get<PaginatedResponse<SceneBoundary>>(
      `${this.basePath}/${detectionId}/boundaries`,
      { params: params as unknown as Record<string, string | number | boolean | undefined> }
    );
  }

  /**
   * Detect scene boundaries only (without full analysis)
   *
   * Requires: scene:detect permission
   */
  async detectBoundaries(
    mediaId: string,
    mediaType: 'video' | 'stream' | 'recording',
    options?: { sensitivity?: number; min_gap?: number }
  ): Promise<SceneBoundary[]> {
    return this.client.post<SceneBoundary[]>(`${this.basePath}/boundaries`, {
      media_id: mediaId,
      media_type: mediaType,
      ...options,
    });
  }


  /**
   * Get shots for a scene
   *
   * Requires: scene:read permission
   */
  async getShots(
    detectionId: string,
    sceneId: string,
    params?: PaginationParams
  ): Promise<PaginatedResponse<Shot>> {
    return this.client.get<PaginatedResponse<Shot>>(
      `${this.basePath}/${detectionId}/scenes/${sceneId}/shots`,
      { params: params as unknown as Record<string, string | number | boolean | undefined> }
    );
  }

  /**
   * Get all shots for a detection
   *
   * Requires: scene:read permission
   */
  async getAllShots(
    detectionId: string,
    params?: PaginationParams & { shot_type?: ShotType }
  ): Promise<PaginatedResponse<Shot & { scene_id: string }>> {
    return this.client.get(`${this.basePath}/${detectionId}/shots`, { params: params as unknown as Record<string, string | number | boolean | undefined> });
  }


  /**
   * Get scene summary/statistics
   *
   * Requires: scene:read permission
   */
  async getSummary(
    detectionId: string
  ): Promise<{
    total_scenes: number;
    total_shots: number;
    scene_types: Record<SceneType, number>;
    shot_types: Record<ShotType, number>;
    average_scene_duration: number;
    average_shot_duration: number;
    dominant_colors: string[];
    content_labels: Array<{ label: string; count: number }>;
  }> {
    return this.client.get(`${this.basePath}/${detectionId}/summary`);
  }

  /**
   * Get visual timeline
   *
   * Requires: scene:read permission
   */
  async getTimeline(
    detectionId: string,
    options?: { resolution?: number; include_shots?: boolean }
  ): Promise<
    Array<{
      timestamp: number;
      scene_id: string;
      scene_type: SceneType;
      shot_id?: string;
      shot_type?: ShotType;
      thumbnail_url?: string;
    }>
  > {
    return this.client.get(`${this.basePath}/${detectionId}/timeline`, {
      params: options,
    });
  }

  /**
   * Compare scenes between detections
   *
   * Requires: scene:read permission
   */
  async compareScenes(
    sourceDetectionId: string,
    targetDetectionId: string,
    options?: { min_similarity?: number }
  ): Promise<SceneComparison[]> {
    return this.client.post<SceneComparison[]>(`${this.basePath}/compare`, {
      source_detection_id: sourceDetectionId,
      target_detection_id: targetDetectionId,
      ...options,
    });
  }

  /**
   * Find similar scenes across all content
   *
   * Requires: scene:read permission
   */
  async findSimilarScenes(
    detectionId: string,
    sceneId: string,
    options?: { limit?: number; min_similarity?: number }
  ): Promise<
    Array<{
      detection_id: string;
      scene_id: string;
      media_id: string;
      similarity: number;
      thumbnail_url?: string;
    }>
  > {
    return this.client.get(`${this.basePath}/${detectionId}/scenes/${sceneId}/similar`, {
      params: options,
    });
  }


  /**
   * Export scene data
   *
   * Requires: scene:read permission
   */
  async exportDetection(
    detectionId: string,
    format: 'json' | 'csv' | 'edl' | 'fcpxml'
  ): Promise<{ url: string; expires_at: string }> {
    return this.client.post(`${this.basePath}/${detectionId}/export`, { format });
  }

  /**
   * Generate scene thumbnails
   *
   * Requires: scene:update permission
   */
  async generateThumbnails(
    detectionId: string,
    options?: { regenerate?: boolean; format?: 'jpg' | 'png'; size?: string }
  ): Promise<{ generated: number }> {
    return this.client.post(`${this.basePath}/${detectionId}/thumbnails`, options);
  }


  /**
   * Wait for scene detection to complete
   */
  async waitForReady(
    detectionId: string,
    options?: {
      pollInterval?: number;
      timeout?: number;
      onProgress?: (detection: SceneDetection) => void;
    }
  ): Promise<SceneDetection> {
    const pollInterval = options?.pollInterval || 3000;
    const timeout = options?.timeout || 1800000; // 30 minutes
    const startTime = Date.now();

    while (Date.now() - startTime < timeout) {
      const detection = await this.getDetection(detectionId);

      if (options?.onProgress) {
        options.onProgress(detection);
      }

      if (detection.status === 'ready') {
        return detection;
      }

      if (detection.status === 'failed') {
        throw new Error(
          `Scene detection failed: ${detection.error || 'Unknown error'}`
        );
      }

      await new Promise((resolve) => setTimeout(resolve, pollInterval));
    }

    throw new Error(`Scene detection timed out after ${timeout}ms`);
  }

  /**
   * Merge scenes
   *
   * Requires: scene:update permission
   */
  async mergeScenes(
    detectionId: string,
    sceneIds: string[],
    options?: { scene_type?: SceneType; description?: string }
  ): Promise<Scene> {
    return this.client.post<Scene>(
      `${this.basePath}/${detectionId}/scenes/merge`,
      { scene_ids: sceneIds, ...options }
    );
  }

  /**
   * Split scene at timestamp
   *
   * Requires: scene:update permission
   */
  async splitScene(
    detectionId: string,
    sceneId: string,
    splitTime: number
  ): Promise<{ first: Scene; second: Scene }> {
    return this.client.post(`${this.basePath}/${detectionId}/scenes/${sceneId}/split`, {
      split_time: splitTime,
    });
  }
}

/**
 * Create a Scene API instance
 */
export function createSceneAPI(client: WaveClient): SceneAPI {
  return new SceneAPI(client);
}

File: sdk-python/wave_sdk/scene.py
"""WAVE SDK - Scene API. AI scene detection, shot classification, and visual analysis."""
from __future__ import annotations
import time
from typing import Any
from pydantic import BaseModel
from wave.client import WaveClient

class Scene(BaseModel):
    id: str; start_time: float; end_time: float; scene_type: str; shot_type: str | None = None; confidence: float; labels: list[str] | None = None; thumbnail_url: str | None = None

class Shot(BaseModel):
    id: str; scene_id: str; start_time: float; end_time: float; type: str; confidence: float

class SceneDetection(BaseModel):
    id: str; asset_id: str; status: str; scene_count: int = 0; shot_count: int = 0; scenes: list[Scene] | None = None; created_at: str; updated_at: str

class SceneAPI:
    def __init__(self, client: WaveClient): self._client = client; self._base = "/v1/scene"
    def detect(self, media_id: str, media_type: str = "video", **kwargs: Any) -> SceneDetection: return SceneDetection(**self._client.post(f"{self._base}/detect", json={"media_id": media_id, "media_type": media_type, **kwargs}))
    def get_detection(self, detection_id: str) -> SceneDetection: return SceneDetection(**self._client.get(f"{self._base}/{detection_id}"))
    def list_detections(self, **params: Any) -> dict: return self._client.get(self._base, params={k: v for k, v in params.items() if v is not None})
    def remove_detection(self, detection_id: str) -> None: self._client.delete(f"{self._base}/{detection_id}")
    def get_scenes(self, detection_id: str) -> list[Scene]: return [Scene(**s) for s in self._client.get(f"{self._base}/{detection_id}/scenes")]
    def get_scene(self, detection_id: str, scene_id: str) -> Scene: return Scene(**self._client.get(f"{self._base}/{detection_id}/scenes/{scene_id}"))
    def update_scene(self, detection_id: str, scene_id: str, **kwargs: Any) -> Scene: return Scene(**self._client.patch(f"{self._base}/{detection_id}/scenes/{scene_id}", json=kwargs))
    def get_scene_at_time(self, detection_id: str, time_seconds: float) -> Scene: return Scene(**self._client.get(f"{self._base}/{detection_id}/at-time", params={"time": time_seconds}))
    def get_all_shots(self, detection_id: str) -> list[Shot]: return [Shot(**s) for s in self._client.get(f"{self._base}/{detection_id}/shots")]
    def get_shots(self, detection_id: str, scene_id: str) -> list[Shot]: return [Shot(**s) for s in self._client.get(f"{self._base}/{detection_id}/scenes/{scene_id}/shots")]
    def detect_boundaries(self, detection_id: str) -> list[dict]: return self._client.get(f"{self._base}/{detection_id}/boundaries")
    def get_boundaries(self, detection_id: str) -> list[dict]: return self._client.get(f"{self._base}/{detection_id}/boundaries")
    def split_scene(self, detection_id: str, scene_id: str, split_time: float) -> list[Scene]: return [Scene(**s) for s in self._client.post(f"{self._base}/{detection_id}/scenes/{scene_id}/split", json={"split_time": split_time})]
    def merge_scenes(self, detection_id: str, scene_ids: list[str]) -> Scene: return Scene(**self._client.post(f"{self._base}/{detection_id}/merge", json={"scene_ids": scene_ids}))
    def find_similar_scenes(self, detection_id: str, scene_id: str, **params: Any) -> list[Scene]: return [Scene(**s) for s in self._client.get(f"{self._base}/{detection_id}/scenes/{scene_id}/similar", params={k: v for k, v in params.items() if v is not None})]
    def compare_scenes(self, detection_ids: list[str]) -> dict: return self._client.post(f"{self._base}/compare", json={"detection_ids": detection_ids})
    def generate_thumbnails(self, detection_id: str) -> SceneDetection: return SceneDetection(**self._client.post(f"{self._base}/{detection_id}/thumbnails"))
    def get_summary(self, detection_id: str) -> dict: return self._client.get(f"{self._base}/{detection_id}/summary")
    def get_timeline(self, detection_id: str) -> dict: return self._client.get(f"{self._base}/{detection_id}/timeline")
    def export_detection(self, detection_id: str, format: str = "json") -> dict: return self._client.get(f"{self._base}/{detection_id}/export", params={"format": format})
    def wait_for_ready(self, detection_id: str, poll_interval: float = 3.0, timeout: float = 600.0) -> SceneDetection:
        start = time.time()
        while time.time() - start < timeout:
            d = self.get_detection(detection_id)
            if d.status == "ready": return d
            if d.status == "failed": raise RuntimeError("Scene detection failed")
            time.sleep(poll_interval)
        raise TimeoutError(f"Scene detection timed out after {timeout}s")
